mlir/lib/Transforms/ViewOpGraph.cpp

```diff
@@ -261,9 +273,18 @@ class PrintOpPass : public impl::ViewOpGraphBase<PrintOpPass> {
         if (printControlFlowEdges && prevNode)
           emitEdgeStmt(*prevNode, nextNode, /*label=*/"",
                        kLineStyleControlFlow);
+
+        // If we are at first iteration, save the first operand node for the
+        // incoming edges.
+        if (not prevNode) {
+          blockFirstNodeMap[&block] = nextNode;
+        }
```

Comment:
Just leave this out of the `for (Operation &op :block)` loop.
Something like:
```c++
if (not block.empty()) {
  blockFirstNodeMap[&block] = &*block.begin();
  blockLastNodeMap[&block] = &*block.end();
}
for (Operation &op : block) {
  // ... as before ...
}
```

Reply:
Eh, there is a little problem with this. The point is that I guess you understood that is sufficient to store a pointer to the `Operation` in those maps, and later use it when emitting the edges. Reality is a little bit more complicated:
The `emitEdgeStmt` directly takes as argument the `Node`, which is a struct representing the node object in DOT language. Therefore, we cannot leave out the storage of the `Node`s in the maps, because we actually need to wait the `processOperation` invocation which returns the `Node` object.

In order to follow your suggestions, we could indeed modify the maps to store the pointers to `Operations`, but we anyway need to later have a way to go from `Operation` to `Node` representing that operation in the DOT, so we are adding another map to keep this link, and we need to populate that map when invoking `processOperation`.

Another idea could be to modify the `emitEdgeStmt` in order to accept as argument the `Operation`s, and stub the `processOperation` to automatically popuplate a map with `Operation` -> `Node`. This however would need a quite heavy modification in the codebase, reducing in my opinion the chance of getting a version of this merged.

Let me know if I explained the problem clearly enough.

Reply:
Oh I see. Let's leave this like it is right now. Thanks